### ISSUE TITLE
Extra "\r" characters in chunked request body

### ISSUE THREAD

**Issue by rkistner** [puma/puma#2325]
**Describe the bug**

When a client is using `Transfer-Encoding: chunked`, in some edge cases, the read body (`env['rack.input'].read`) will contain extra `\r` characters.

This happens when:
 * The body contains HTTP chunks larger than 4096 bytes in size.
 * One of these end exactly 1 byte before a chunked read boundary, splitting the `\r\n` delimiter in two.

This means it's fairly rare to happen in typical cases, but it is effectively silent data corruption when it does happen.

**Puma config:**

The default.

**To Reproduce**

Pull request with a failing unit test will follow soon.

**Expected behavior**

The body should be parsed correctly.

**Desktop:**

 - OS: Tested on Linux and MacOS.
 - Puma Version:  Tested on 4.2.1 and current master.

**Comment by nateberkopec** [comment]
Interesting. We'll have to be careful with the fix here, as chunked encoding + line endings are both bug/security hotspots.

**Comment by rkistner** [comment]
I'm fairly confident that my change fixes this issue, and doesn't introduce any regressions cases. Of course, it would be good for a maintainer to confirm this.

I've also found:
1. There are currently no checks that a chunk ends with `\r\n`, the last two characters are simply stripped.
2. The `decode_chunk` method does currently have many edge-case checks, and it's difficult to prove that it does not contain more bugs. It would be good if the logic could be simplified, but I'm not sure how feasible that is without decreasing performance.

**Comment by wjessop** [comment]
I think I just hit something similar to this, but with tiny body sizes. Making a POST request with a JSON body in Go (which uses chunked encoding by default unless you supply a content length) failed. Specifying a content-length to the request from Go switches chunked encoding off, and these non-chunked requests worked fine. Here's the Go requests as seen by Netcat:

**Chunked, doesn't work against Puma**

```
(master) onion:web-app will$ nc -l 3000
POST /api/v2/users/sign_in HTTP/1.1
Host: localhost:3000
User-Agent: Go-http-client/1.1
Transfer-Encoding: chunked
Accept: application/json
Content-Type: application/json; charset=utf-8
Accept-Encoding: gzip
Connection: close

4e
{"user":{"email":"user0@email.com","password":"Password0","provider":"email"}}
0

^C
```

Here's the log output of the failed request, notice the lack of params:

```
2020-08-01 20:34:30.729379 I [9101:puma threadpool 001] {ip: ::1} (1.271s) Api::V2::SessionsController -- Completed #create -- { :controller => "Api::V2::SessionsController", :action => "create", :format => "JSON", :method => "POST", :path => "/api/v2/users/sign_in", :view_runtime => 0.0, :db_runtime => 0.0, :status => 401, :status_message => "Unauthorized" }
```

Here's what the request body is when I put byebug in there:

```
(byebug) JSON.parse(request.body.read)
{"user"=>{"email"=>"user0@email.com", "password"=>"Password0", "provider"=>"email"}}
(byebug) 
```

**Un-chunked, works fine**

```
(master) onion:web-app will$ nc -l 3000
POST /api/v2/users/sign_in HTTP/1.1
Host: localhost:3000
User-Agent: Go-http-client/1.1
Content-Length: 78
Accept: application/json
Content-Type: application/json; charset=utf-8
Accept-Encoding: gzip
Connection: close

{"user":{"email":"user0@email.com","password":"Password0","provider":"email"}}^C
```

Here's the log output of the identical, but un-chunked request, notice that there are params here this time:

```
2020-08-01 21:11:44.776350 I [23204:puma threadpool 001] {ip: ::1} (290.1ms) Api::V2::SessionsController -- Completed #create -- { :controller => "Api::V2::SessionsController", :action => "create", :params => { "user" => { "email" => "user0@email.com", "password" => "[FILTERED]", "provider" => "email" } }, :format => "JSON", :method => "POST", :path => "/api/v2/users/sign_in", :status => 200, :view_runtime => 20.58, :db_runtime => 18.3, :status_message => "OK" }

```

I'm not sure if this is related, if not I can create a separate issue.

**Comment by rkistner** [comment]
@wjessop That looks like a different issue. This one is only triggered with larger payloads (20kb minimum), and corrupts te data in `request.body.read`. In your case the request data seems fine, just not automatically parsed as parameters.

**Comment by nateberkopec** [comment]
Fixed in #2236
